File: docusewa/docusewa-web/src/services/service-request.service.ts
import { createClient } from '@/lib/supabase';
import type {
  ServiceRequest,
  CreateServiceRequestInput,
  ServiceResult,
  RequestStatus,
} from '@/types/janseva';

// ---------------------------------------------------------------------------
// JanSeva Service Request Service
//
// RLS automatically scopes all queries to the authenticated citizen.
// A citizen created via Flutter and accessing via web gets the SAME requests.
// ---------------------------------------------------------------------------

/**
 * Fetch all service requests for the authenticated citizen.
 * Sorted newest first. RLS ensures only own requests are returned.
 */
export async function getRequests(): Promise<ServiceResult<ServiceRequest[]>> {
  const supabase = createClient();

  const { data, error } = await supabase
    .from('service_requests')
    .select('*')
    .order('created_at', { ascending: false });

  if (error) {
    return { data: null, error: error.message };
  }

  return { data: data as ServiceRequest[], error: null };
}

/**
 * Fetch a single service request by ID (RLS ensures ownership).
 */
export async function getRequestById(
  id: string
): Promise<ServiceResult<ServiceRequest>> {
  const supabase = createClient();

  const { data, error } = await supabase
    .from('service_requests')
    .select('*')
    .eq('id', id)
    .single();

  if (error) {
    return { data: null, error: error.message };
  }

  return { data: data as ServiceRequest, error: null };
}

/**
 * Create a new service request.
 * user_id is set from auth.uid() by RLS policy (insert check).
 */
export async function createRequest(
  input: CreateServiceRequestInput
): Promise<ServiceResult<ServiceRequest>> {
  const supabase = createClient();

  const {
    data: { user },
  } = await supabase.auth.getUser();

  if (!user) {
    return { data: null, error: 'Not authenticated.' };
  }

  const { data, error } = await supabase
    .from('service_requests')
    .insert({
      user_id: user.id,          // enforced by RLS insert check too
      title: input.title,
      description: input.description ?? null,
      category: input.category,
      priority: input.priority ?? 'normal',
      status: 'submitted',       // new requests go straight to submitted
    })
    .select()
    .single();

  if (error) {
    return { data: null, error: error.message };
  }

  return { data: data as ServiceRequest, error: null };
}

/**
 * Cancel a citizen's own request (status must not already be terminal).
 */
export async function cancelRequest(
  id: string
): Promise<ServiceResult<null>> {
  const supabase = createClient();

  const { error } = await supabase
    .from('service_requests')
    .update({ status: 'cancelled' as RequestStatus })
    .eq('id', id)
    .in('status', ['draft', 'submitted', 'under_review', 'additional_info_required']);

  if (error) {
    return { data: null, error: error.message };
  }

  return { data: null, error: null };
}

/**
 * Subscribe to real-time status changes on the citizen's requests.
 * Works across web and mobile — same Supabase channel.
 */
export function subscribeToRequestUpdates(
  userId: string,
  onUpdate: (request: ServiceRequest) => void
) {
  const supabase = createClient();

  const channel = supabase
    .channel(`service_requests_${userId}`)
    .on(
      'postgres_changes',
      {
        event: '*',
        schema: 'public',
        table: 'service_requests',
        filter: `user_id=eq.${userId}`,
      },
      (payload) => {
        if (payload.new) {
          onUpdate(payload.new as ServiceRequest);
        }
      }
    )
    .subscribe();

  return () => {
    supabase.removeChannel(channel);
  };
}

File: docseva/lib/services/service_request_service.dart
import 'package:docusewa/core/supabase_client.dart';
import 'package:docusewa/models/service_request.dart';
import 'package:supabase_flutter/supabase_flutter.dart';

/// DocuSewa Service Request Service.
/// RLS auto-scopes all queries to the authenticated citizen.
class ServiceRequestService {
  ServiceRequestService._();
  static final ServiceRequestService instance = ServiceRequestService._();

  /// Fetch all service requests for the authenticated citizen, newest first.
  Future<List<ServiceRequest>> getRequests() async {
    try {
      final data = await DocuSewaSupabase.client
          .from('service_requests')
          .select()
          .order('created_at', ascending: false);
      return (data as List).map((e) => ServiceRequest.fromJson(e)).toList();
    } catch (_) {
      return [];
    }
  }

  /// Fetch a single service request by ID (RLS ensures ownership).
  Future<ServiceRequest?> getRequestById(String id) async {
    try {
      final data = await DocuSewaSupabase.client
          .from('service_requests')
          .select()
          .eq('id', id)
          .single();
      return ServiceRequest.fromJson(data);
    } catch (_) {
      return null;
    }
  }

  /// Create a new service request.
  Future<ServiceRequest?> createRequest({
    required String title,
    String? description,
    required ServiceCategory category,
    RequestPriority priority = RequestPriority.normal,
  }) async {
    try {
      final userId = DocuSewaSupabase.currentUserId;
      if (userId == null) return null;

      final data = await DocuSewaSupabase.client
          .from('service_requests')
          .insert({
            'user_id': userId,
            'title': title,
            'description': description,
            'category': category.value,
            'priority': priority.value,
            'status': 'submitted',
          })
          .select()
          .single();
      return ServiceRequest.fromJson(data);
    } catch (_) {
      return null;
    }
  }

  /// Cancel a request (only non-terminal requests can be cancelled).
  Future<bool> cancelRequest(String id) async {
    try {
      await DocuSewaSupabase.client
          .from('service_requests')
          .update({'status': 'cancelled'})
          .eq('id', id)
          .inFilter('status', ['draft', 'submitted', 'under_review', 'additional_info_required']);
      return true;
    } catch (_) {
      return false;
    }
  }

  /// Subscribe to real-time updates on the citizen's requests.
  /// Returns an unsubscribe function.
  Future<void> Function() subscribeToRequests(
    String userId,
    void Function(ServiceRequest) onUpdate,
  ) {
    final channel = DocuSewaSupabase.client
        .channel('service_requests_$userId')
        .onPostgresChanges(
          event: PostgresChangeEvent.all,
          schema: 'public',
          table: 'service_requests',
          filter: PostgresChangeFilter(
            type: PostgresChangeFilterType.eq,
            column: 'user_id',
            value: userId,
          ),
          callback: (payload) {
            try {
              if (payload.newRecord.isNotEmpty) {
                onUpdate(ServiceRequest.fromJson(payload.newRecord));
              }
            } catch (_) {}
          },
        )
        .subscribe();

    return () async {
      await DocuSewaSupabase.client.removeChannel(channel);
    };
  }
}
